Rewrite content about how the app manager processes helm charts (#1045)
replicatedhq/replicated-docs: docs/vendor/helm-overview.md

    ```yaml
    apiVersion: kustomize.config.k8s.io/v1beta1
    bases:
    - ../../../../midstream/charts/postgresql
    kind: Kustomization
    ```
   
   End users can edit the downstream Kustomization files to make changes before deploying the application. Any instructions that users add to the Kustomization files in the downstream directory take priority over midstream and base Kustomization files. For more information about how users can make changes before deploying, see [Patching with Kustomize](/enterprise/updating-patching-with-kustomize) in _Enterprise User Documentation_.

1. **Deploys the Helm chart**: The app manager runs `kustomize build` for any Kustomization files in the `overlays/downstream/charts` directory. The app manager then packages the resulting manifests into a new tarball for Helm to consume.

   Finally, the app manager runs `helm upgrade -i chart.tar.gz`. The Helm binary processes hooks and weights, applies manifests to the Kubernetes cluster, and saves a release secret similar to `sh.helm.release.v1.chart-name.v1`. Helm uses this secret to track upgrades and rollbacks of applications.
#### Replicated Helm

> :::note

Reviewer: I suggest turning this note into a partial bc there are a couple of places that it can be reused.

Reviewer: P.S. It's good content!

Deploying an application with the helm CLI differs from the _native Helm_ deployment method described above because, when users install with the helm CLI directly, they have access to all Helm functionality. Some enterprise users prefer or require using the helm CLI because their existing CI/CD pipeline is already compatible with Helm charts. Similarly, enterprise users might have organizational policies that require using Helm to manage applications.

Users do not have access to certain Replicated features when they install and manage the application with the helm CLI directly. This is because the app manager does not manage the lifecycle of the application. For example, users must update the application using the `helm upgrade` command, rather than using the admin console UI or the kots CLI.

For more information about how to package an application with Replicated so that users can install using the helm CLI, see [Supporting helm CLI Installations (Beta)](helm-install).

### Air Gap

> The app manager supports Helm installations into air gap environments. When a user installs a Helm chart-based application in an air gap environment, the chart processing is managed in the end user environment. This means that the app manager can use user-supplied values, license values, and existing values to create deployable manifests.

Reviewer: I think just native Helm, since a note comes at the end about Helm CLI. But I think that should be specified here to avoid confusion - and also to help steer folks away from Replicated Helm.

> The app manager supports Helm installations into air gap environments. When a user installs a Helm chart-based application in an air gap environment, the chart processing is managed in the end user environment. This means that the app manager can use user-supplied values, license values, and existing values to create deployable manifests.

Author: This is a good question... will follow up with the Orange team for confirmation

> The app manager supports Helm installations into air gap environments. When a user installs a Helm chart-based application in an air gap environment, the chart processing is managed in the end user environment. This means that the app manager can use user-supplied values, license values, and existing values to create deployable manifests.

Author: @ajp-io would you be able to confirm if air gap installation are supported for native Helm and Replicated Helm? I didn't see anything in the docs about the `builder` attribute that suggests it's limited to native Helm, but we weren't sure.

> The app manager supports Helm installations into air gap environments. When a user installs a Helm chart-based application in an air gap environment, the chart processing is managed in the end user environment. This means that the app manager can use user-supplied values, license values, and existing values to create deployable manifests.

Reviewer 2: @paigecalvert Airgap installs are supported for both native Helm and Replicated Helm.

To create an `.airgap` bundle for a release that uses Helm charts, the Replicated vendor portal renders templates of the Helm charts with `helm template`. To specify which values from the Helm chart `values.yaml` file are included in the `.airgap` bundle, you add a `builder` key in the HelmChart custom resource manifest file. For more information, see [builder](/reference/custom-resource-helmchart#builder) in _HelmChart_.

:::note
The helm CLI installation method does not support installations into air gap environments. See [helm CLI Limitations](#helm-cli-limitations) below.
:::

## Limitations

This section lists the limitations for the Helm chart installation methods.

There are different limitations depending on if your customers install and manage the application with the app manager or if they use the helm CLI directly:

> * [Replicated Helm and Native Helm Limitations](#replicated-helm-limitations)

Reviewer: Suggest switching the order to "Native Helm and Replicated Helm Limitations" to match the heading that follows.

  The following screenshot from the Replicated admin console shows a `base/charts/` directory for a deployed application. The `base/charts/` directory contains a Helm chart named postgresql with one subchart:

  ![Base directory in the admin console UI](/images/native-helm-base.png)

  In the screenshot above, a Kustomization file that targets the resources from the postgresql Helm chart appears in the `base/charts/postgresql/` directory:

   ```yaml
   apiVersion: kustomize.config.k8s.io/v1beta1
   kind: Kustomization
   resources:
   - secrets.yaml
   - statefulset.yaml
   - svc-headless.yaml
   - svc.yaml
   ```

1. **Writes midstream files with Kustomize instructions from Replicated**: The app manager then copies the directory structure from `base/charts/` to an `overlays/midstream/charts/` directory. The following screenshot shows an example of the midstream directory for the postgresql Helm chart: 
   
  ![Midstream directory in the admin console UI](/images/native-helm-midstream.png)

>   As shown in the screenshot above, the midstream directory also contains a Kustomization file with instructions from Replicated for all deployed resources, such as imagePullSecrets, image proxies, and backup labels. For example, in the midstream Kustomization file, the app manager renames any private images to point to the Replicated proxy service.

Reviewer 2: Some small suggestions for this piece
```suggestion
  As shown in the screenshot above, the midstream directory also contains a Kustomization file with instructions from Replicated for all deployed resources, such as image pull secrets, image re-writes, and backup labels. For example, in the midstream Kustomization file, the app manager re-writes any private images to pull from the Replicated proxy service.
```

   ![Downstream directory in the admin console UI](/images/native-helm-downstream.png)

   As shown in the screenshot above, each chart and subchart directory in the downstream directory also contains a Kustomization file. These downstream Kustomization files contain only a `bases` entry that references the corresponding Kustomization file from the midstream directory. For example:

    ```yaml
    apiVersion: kustomize.config.k8s.io/v1beta1
    bases:
    - ../../../../midstream/charts/postgresql
    kind: Kustomization
    ```
   
   End users can edit the downstream Kustomization files to make changes before deploying the application. Any instructions that users add to the Kustomization files in the downstream directory take priority over midstream and base Kustomization files. For more information about how users can make changes before deploying, see [Patching with Kustomize](/enterprise/updating-patching-with-kustomize) in _Enterprise User Documentation_.

> 1. **Deploys the Helm chart**: The app manager runs `kustomize build` for any Kustomization files in the `overlays/downstream/charts` directory. The app manager then packages the resulting manifests into a new tarball for Helm to consume.

Reviewer 2: Small edit for accuracy.  We don't actually tarball the new chart before Helm consumes it.

```suggestion
1. **Deploys the Helm chart**: The app manager runs `kustomize build` for any Kustomization files in the `overlays/downstream/charts` directory. The app manager then packages the resulting manifests into a new chart for Helm to consume.
```

> 1. **Deploys the Helm chart**: The app manager runs `kustomize build` for any Kustomization files in the `overlays/downstream/charts` directory. The app manager then packages the resulting manifests into a new tarball for Helm to consume.
> 
>    Finally, the app manager runs `helm upgrade -i chart.tar.gz`. The Helm binary processes hooks and weights, applies manifests to the Kubernetes cluster, and saves a release secret similar to `sh.helm.release.v1.chart-name.v1`. Helm uses this secret to track upgrades and rollbacks of applications.

Reviewer 2: Small edit for accuracy.  We don't actually tarball the new chart before Helm consumes it.

```suggestion
1. **Deploys the Helm chart**: The app manager runs `kustomize build` for any Kustomization files in the `overlays/downstream/charts` directory. The app manager then packages the resulting manifests into a new chart for Helm to consume.

   Finally, the app manager runs `helm upgrade -i`. The Helm binary processes hooks and weights, applies manifests to the Kubernetes cluster, and saves a release secret similar to `sh.helm.release.v1.chart-name.v1`. Helm uses this secret to track upgrades and rollbacks of applications.
```

Reviewer 2: Also edited this to include the default flags `--timeout 3600s -n <namespace>`

#### Replicated Helm

:::note
<ReplicatedHelmDeprecated/>
:::

With the Replicated Helm deployment method, the app manager renders the Helm templates and deploys them as standard Kubernetes manifests using `kubectl apply`. The app manager also has additional functionality for specific Helm hooks. For example, when the app manager encounters an upstream Helm chart with a `helm.sh/hook-delete-policy` annotation, it automatically adds the same `kots.io/hook-delete-policy` to the Job object.

> To deploy charts with the Replicated Helm method, the app manager runs `helm upgrade -i chart.tar.gz --timeout 60m -n <namespace>`. The Helm binary processes hooks and weights, applies manifests to the Kubernetes cluster, and saves a release secret similar to `sh.helm.release.v1.chart-name.v1`. Helm uses this secret to track upgrades and rollbacks of applications.

Reviewer 2: This looks like it was supposed to be in the native Helm section.
```suggestion
```

Replicated supports delivering an enterprise application as Helm charts, or including Helm charts as components of an application. An application can use more than one Helm chart, and can use more than a single instance of any Helm chart.

To deploy a Helm chart, start by adding the Helm chart to a release in the Replicated vendor portal. For information about how to create a new release from an existing Helm chart, see [Creating Releases with Helm Charts](helm-release).

## Helm Chart Installation Options

When you distribute an application packaged with Helm, your users can install and manage the application with either the Replicated app manager or the helm CLI.

This section describes the processes for installing Helm charts with the app manager or the helm CLI.
### App Manager

Users can install an application packaged with Helm charts using the app manager in either an existing cluster or a cluster provisioned by the Replicated Kubernetes installer. The app manager also supports Helm installations into air gap environments. For more information, see [Air Gap](#air-gap) below.

> As an application vendor, you specify whether the app manager uses the _native Helm_ or _Replicated Helm_ method to deploy your Helm chart-based application. You specify the deployment method in the `useHelmInstall` field of the Replicated HelmChart custom resource manifest file for the Helm chart. For more information, see [useHelmInstall](/reference/custom-resource-helmchart#usehelminstall) in _HelmChart_.

Reviewer: I'm not sure if I thought to mention this before, but maybe we ought to add something like "The Replicated Helm method is deprecated and can only be used for iterating on existing Replicated Helm releases." WDYT?

Reviewer: Nevermind, I see you added a note later on. :)

To create an `.airgap` bundle for a release that uses Helm charts, the Replicated vendor portal renders templates of the Helm charts with `helm template`. To specify which values from the Helm chart `values.yaml` file are included in the `.airgap` bundle, you add a `builder` key in the HelmChart custom resource manifest file. For more information, see [builder](/reference/custom-resource-helmchart#builder) in _HelmChart_.

:::note
The helm CLI installation method does not support installations into air gap environments. See [helm CLI Limitations](#helm-cli-limitations) below.
:::

## Limitations

This section lists the limitations for the Helm chart installation methods.

There are different limitations depending on if your customers install and manage the application with the app manager or if they use the helm CLI directly:

* [Native Helm and Replicated Helm Limitations](#replicated-helm-limitations)
* [helm CLI Limitations](#helm-cli-limitations)

### Native Helm and Replicated Helm Limitations {#replicated-helm-limitations}

The following limitations apply when using the app manager to install and manage Helm charts:

* <ReplicatedHelmDeprecated/>

> * <TemplateLimitation/>

Author: thank you for catching that!
